kernel/futex.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2019 Intel corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <spinlock.h>
+#include <kswap.h>
+#include <wait_q.h>
+#include <syscall_handler.h>
+#include <init.h>
+#include <ksched.h>
+
+static struct k_spinlock lock;
+
+static _wait_q_t *k_futex_find_waitq(struct k_futex *futex)
+{
+	struct _k_object *obj;
+
+	obj = z_object_find(futex);
+	if (obj == NULL || obj->type != K_OBJ_FUTEX) {
+		return NULL;
+	}
+
+	return (_wait_q_t *)obj->data;
+}
+
+void z_impl_k_futex_init(struct k_futex *futex)
+{
+	futex->val = 0U;
+	z_object_init(futex);
+}
+
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(k_futex_init, futex)
+{
+	if (Z_SYSCALL_MEMORY_WRITE(futex, sizeof(struct k_futex)) != 0) {
+		return -EACCES;
+	}
+
+	z_impl_k_futex_init((struct k_futex *)futex);
+
+	return 0;
+}
+#endif
+
+int z_impl_k_futex_wake(struct k_futex *futex, unsigned int num_waiters)
+{
+	k_spinlock_key_t key;
+	_wait_q_t *thread_q;
+	struct k_thread *thread;
+	unsigned int woken = 0;
+
+	thread_q = k_futex_find_waitq(futex);
+	if (thread_q == NULL) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&lock);
+
+	while (woken < num_waiters &&
+	    (thread = z_unpend_first_thread(thread_q))) {
+		z_ready_thread(thread);
+		z_set_thread_return_value(thread, 0);
+		woken++;
+	}
+
+	if (woken) {
+		z_reschedule(&lock, key);
+	} else {
+		k_spin_unlock(&lock, key);
+	}
+
+	return woken;
+}
+
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(k_futex_wake, futex, num_waiters)
+{
+	if (Z_SYSCALL_MEMORY_WRITE(futex, sizeof(struct k_futex)) != 0) {
+		return -EACCES;
+	}
+
+	return z_impl_k_futex_wake((struct k_futex *)futex, num_waiters);
+}
+#endif
+
+int z_impl_k_futex_wait(struct k_futex *futex, int expected, s32_t timeout)
+{
+	k_spinlock_key_t key;
+	_wait_q_t *thread_q;
+	int ret;
+
+	thread_q = k_futex_find_waitq(futex);
+	if (thread_q == NULL) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&lock);
+
+	if (atomic_get(&futex->val) != (atomic_val_t)expected) {
+		k_spin_unlock(&lock, key);
+		return -EAGAIN;
+	}
+
+	if (timeout == K_NO_WAIT) {
+		k_spin_unlock(&lock, key);
+		return -EBUSY;
+	}
+
+	ret = z_pend_curr(&lock, key, thread_q, timeout);
+	if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
```

Comment:
I see what you're doing here: you want separate return values for "we raced on the expected value, try again" and "the timeout expired".

Nonetheless this is confusing.  In fact everywhere else in zephyr where a timeout expires, the relevant call returns -EAGAIN.    In fact this is the default behavior of the pend API in the scheduler, for exactly this reason.  We should be consistent here.  I know the error says "timed out", but that's not the error we use for timeouts, basically. :)

If you need to disambiguate the cases, maybe returning -EINVAL for the "expected has changed" case would be a better choice.  The argument is, in fact, "invalid" for the case where we want to wait.

Reply:
I asked Wentong to return -ETIMEDOUT because to match the return values from Linux.

```
       ETIMEDOUT
              The  operation in futex_op employed the timeout specified in timeout, and the timeout
              expired before the operation completed.
```

> We should be consistent here. I know the error says "timed out", but that's not the error we use for timeouts, basically. :)

Why is this the case? Why don't these APIs return -ETIMEDOUT? We can break APIs for 2.0, maybe we should do that?

I'm trying to keep the return values consistent with Linux to make it simple to port futex-using userspace code to Zephyr.

Reply:
I wouldn't cry if we changed the timeout return.  I agree it's a cleaner naming.  It would be a ton of code to change in all the tests though.  Not it. :)